### ISSUE TITLE
Multi tenency DI

### ISSUE THREAD

**Issue by stephenlautier** [dotnet/orleans#4580]
We are looking into implementing multi-tenancy with DI, do you have any recommendations? Is possible?

What we like to achieve
 
 - Replace some services based on the tenant (which we will handle ofc)
 - When Grain is initiated it will have the specified tenant so constructor injection will work

I just had did this for aspnetcore by using [Grace](https://github.com/ipjohnson/Grace/) (DI container) which supports child containers.

 Basically what I had to do in order to integrate in aspnet is:
 - Implement custom middleware (which most of the time should be one of the first one)
 - Implement some kind of Tenant resolver, which in our case would resolve via domain
 - Replace aspnet `HttpContext.RequestService` with the our child container.

As far as i know in Orleans its not possible to implement something as I did in aspnet.

**Comment by ReubenBond** [comment]
How do you plan to determine the tenant, through `RequestContext`?

I think this is achievable, but I would not call it a supported scenario. If I was trying to create a proof-of-concept for this sort of thing, I would be looking to replace the implementation of `IGrainActivator` in the main container. You can see the default implementation [here](https://github.com/dotnet/orleans/blob/2380adfbfd7cb65e7e946eaaa3a2861935cd30d2/src/Orleans.Runtime/Catalog/DefaultGrainActivator.cs#L11)

I hope that helps, and while it's not *supported*, we are (of course) still happy to answer an questions you might have.

**Comment by stephenlautier** [comment]
Yes, I'm still not 100% sure but I was going to try to implement it using the `RequestContext`. By supported I was more referring whether its possible or not.

I didn't have much time yet to try it out but in the coming days, I will be working on it.

Thank you for your guidance, that will be a good start. I will also try to give it a shot with how the Grace author suggested, which is to reuse the existing scope and use a factory instead which should be faster and might not need to replace anything.

I will keep this update it, thanks again!

**Comment by stephenlautier** [comment]
@ReubenBond I managed to get something working as you had suggested with the `GrainActivator` + Grace 🎉

I had try to use the `RequestContext` however I didnt manage to get it working

```cs
public static IHeroCollectionGrain GetHeroCollectionGrain(this IGrainFactory factory, string tenant)
{
	RequestContext.Set("tenant", tenant);
	return factory.GetGrain<IHeroCollectionGrain>($"tenant\\{tenant}");
}

// within Activator create
var t = RequestContext.Get("tenant"); // always returns null (during StartupTask at least)
```

Currently, my solution is to parse the primary key and take it from there; if i can avoid parsing it and pass it as an item such as `RequestContext` that would be much better if you can help me on that

Another problem I had btw is trying to implement `IGrainActivator`, however `ConstructorArgumentFactory` is `internal`, however, I ended up extending `DefaultGrainActivator` which ended up better for my case.

It still needs a bit more testing to be honest but at least I got it a PoC working 🎆

**Comment by ReubenBond** [comment]
Very glad that it worked for you :) Regarding RequestContext, I see that it is imported just before OnActivateAsync, but not before creating the grain instance (which happens on another context)

**Comment by stephenlautier** [comment]
Sorry, I didnt understand you regarding the `RequestContext`, am I using it wrong or its not possible as I need it?

Pretty much this is how I used it
 - [RequestContext.Get usage](https://github.com/sketch7/orleans-heroes/blob/feature/multi-tenancy/Heroes.SiloHost.ConsoleApp/Infrastructure/TenantGrainActivator.cs#L26)
 -  [RequestContext.Set usage](https://github.com/sketch7/orleans-heroes/blob/feature/multi-tenancy/Heroes.SiloHost.ConsoleApp/WarmupStartupTask.cs#L34) ([definition here](https://github.com/sketch7/orleans-heroes/blob/feature/multi-tenancy/Heroes.Contracts.Grains/GrainFactoryExtensions.cs#L18))

Even tho im not entirely sure with the `RequestContext` in this case, since for example looping and using `RequestContext.Set` wouldnt probably work either, since I guess it will simply replace the same one (as im doing in the Set usage link).

Best way if there was something like the following:

```cs
// pseudo

// get grain + pass activationData
factory.GetGrain<IHeroCollectionGrain>($"tenant\\{tenant}", new Dictionary<object, object> {
  ["tenant"] = tenant
});

// TenantGrainActivator.cs
public override object Create(IGrainActivationContext context)
{
    var tenant = (string)context.Data["tenant"];
    ...
}
```
And extra nice if that would flow to the Grain itself, as currently we parse PrimaryKeys to get the data from, similar to a URI (route)

There seems to be something similar which is `IGrainActivationContext.Items`, however it doesnt seem to be what i want.

**Comment by ReubenBond** [comment]
Interesting idea, @stephenlautier :) This is similar to the discussions about grain identity https://github.com/dotnet/orleans/issues/1123

Do you consider those context bits a part of the grain's identity (i.e, different dictionary values = different grain)?

**Comment by stephenlautier** [comment]
No, I wasn't considering them part of the grains identity; but I guess it can since its metadata of the activation, unless it gets complicated.

**Comment by ReubenBond** [comment]
@stephenlautier apologies for the lack of communication here. Did you manage to get multi-tenancy working satisfactorily?

**Comment by stephenlautier** [comment]
No problem, well we do have a working solution currently, everything seems to be working so far.

With the suggestion above it would have saved me a lot of work, and perhaps more efficient but other than that so far seems fine 👍

**Comment by stephenlautier** [comment]
Sure, thanks for your help
